### DISCUSSION TITLE
Using tRPC in front of an (OpenAPI) REST API

### DISCUSSION THREAD

**Discussion by CoenWarmer** [trpc/trpc#2212] [Q&A] [answered]
I am new to tRPC and I've been spending some time looking over the docs and the related projects like trpc-openapi, but can't quite figure this out, so bear with me.

I have a REST API which is written in Kotlin by a different team, but I would love to use the endpoints it offers in a tRPC setup with a React client.

I could of course write trpc server routes for all of the REST API endpoints and create the zod schema's for the DTOs the REST API provides, but that would be a lot of manual work (I suppose I can convert the DTOs to Zod schema's automatically 🤔, but that still leaves the routes manually)

So I am wondering, would it be possible to give tRPC an OpenAPI spec, and then have it act as a proxy, automatically generating routes + zod schemas..?

**Comment by jlalmes** [comment]
tRPC can't really help you here. For REST API -> tRPC (like) JS client you're best using a tool like: https://www.npmjs.com/package/openapi-client-axios

**Reply by CoenWarmer** [reply]
Yeah, we've been using [OpenAPIGenerator](https://www.npmjs.com/package/@openapitools/openapi-generator-cli) with the typescript-axios client generator. But it still means we have to write react query hooks separately. Using tRPC would effectively solve this.

Out of interest: is this something tRPC can't do _yet_? Or is it not feasible to implement?

**Reply by jlalmes** [reply]
Maybe take a look at [`react-query-swagger`](https://github.com/shaddix/react-query-swagger)? I'm certain that introducing a BFF with `tRPC` is unnecessary overhead unless you have other requirements to do so.

**Reply by CoenWarmer** [reply]
That's amazing, didn't know that existed. Thanks!

**Comment by KATT** [comment, marked as answer]
It's probably not that hard to do. 

A lot of ppl use trpc as a BFF in the same way you're describing by aggregating different microservices and stuff.

**Reply by CoenWarmer** [reply]
Would be a great feature to have (if the tRPC routes could be generated automatically based on OAS I mean).

**Reply by KATT** [reply]
I won't build it as I won't use it, but definitely worth an RFC where ppl could contribute

**Reply by CoenWarmer** [reply]
Will contribute :) love the project!
